src/main/java/org/testng/DependencyMap.java

```diff
@@ -82,13 +83,22 @@ private static boolean hasInstance(
       ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
     Object baseInstance = baseClassMethod.getInstance();
     Object derivedInstance = derivedClassMethod.getInstance();
-    return derivedInstance != null || baseInstance != null;
+    boolean result = derivedInstance != null || baseInstance != null;
+    boolean params = null != baseClassMethod.getFactoryMethodParamsInfo() && null != derivedClassMethod.getFactoryMethodParamsInfo().getParameters();
+    if (result && params && System.getProperty("testng.thread.affinity", "false").equals("true")) {
```

Comment:
Please use `org.testng.internal.RuntimeBehavior#enforceThreadAffinity` instead of querying the JVM argument